### PR TITLE
New PR View / PR Diffs in Visual Studio

### PR DESCRIPTION
**Pull request by CalvinAllen**
This (hopefully) resolves NR-2465 AND NR-38180

High level - all files being diffed are "temp". When we send a message across the wire, they are intercepted and the uri of the temp file is replaced with a `codestream-diff` schemed uri.

<sup> Created from VS Code using [CodeStream](https://codestream.com/?utm_source=cs&utm_medium=pr&utm_campaign=github*com)</sup>

### CODE REVIEW

File: vs/src/CodeStream.VisualStudio.Shared/Services/DotNetBrowserService.cs [hunk 1]
```diff
@@ -402,8 +412,36 @@ public class DotNetBrowserService : IBrowserService, IDisposable, DialogHandler,
...
+		{
+			var messageToken = message?.ToJToken();
+
+			var uriTokens = messageToken?
+				.SelectTokens("$..uri")?
+				.ToList() ?? new List<JToken>();
+		
+			var hasTempFiles = uriTokens
+					.Where(x => x is JValue)
+					.Any(x => x.Value<string>().IsTempFile());
```
Comment (bcanzanella) [inline vs/src/CodeStream.VisualStudio.Shared/Services/DotNetBrowserService.cs:425]: wrap these lines up into a reusable function?

File: vs/src/CodeStream.VisualStudio.Shared/Services/IdeService.cs [hunk 2]
```diff
@@ -513,91 +435,206 @@ public class IdeService : IIdeService {
...
+						null,
+						null,
+						(uint)grfDiffOptions
+					);
+				}
+
+				var diffViewer = GetDiffViewer(frame);
+				diffViewer.Properties.AddProperty(PropertyNames.IsDiff, true);
+				diffViewer.Properties.AddProperty(PropertyNames.IsPRDiff, true);
+				diffViewer.Properties.AddProperty(PropertyNames.OverrideFileUri, rightData.Uri.ToString());
```
Comment (bcanzanella) [inline vs/src/CodeStream.VisualStudio.Shared/Services/IdeService.cs:507]: nice!

File: vs/src/CodeStream.VisualStudio.Shared/WebViewRouter.cs [hunk 3]
```diff
@@ -81,15 +83,39 @@ public class WebViewRouter {
...
 					var target = message.Target();
+
+					var uriTokens = message
+						.Params?
+						.SelectTokens("$..uri")?
+						.ToList() ?? new List<JToken>();
+
+					var hasTempFiles = uriTokens
+						.Where(x => x is JValue)
+						.Any(x => x.Value<string>().IsTempFile());
```
Comment (bcanzanella) [inline vs/src/CodeStream.VisualStudio.Shared/WebViewRouter.cs:94]: same comment

**Review by bcanzanella — CHANGES_REQUESTED** [review]
couple small things

**Comment by CalvinAllen** [comment]
@bcanzanella Ready for another check - moved out to reusable service

**Comment by CalvinAllen** [comment]
@bcanzanella Just noticed a couple extra error handling / null checks things made it into this commit for CLM

**Review by bcanzanella — APPROVED** [review]

**Pull request merged**